File: src/main/java/studio/styx/schemaEXtended/core/schemas/BooleanSchema.java
package studio.styx.schemaEXtended.core.schemas;

import studio.styx.schemaEXtended.core.ParseResult;
import studio.styx.schemaEXtended.core.Schema;

import java.util.List;

public class BooleanSchema extends Schema<Boolean> {
    private String parseError = "The provided value is not a boolean";
    private Boolean defaultValue;

    public BooleanSchema() {}
    public BooleanSchema(String errorMsg) { this.parseError = errorMsg; }

    public BooleanSchema defaultValue(Boolean defaultValue) {
        this.defaultValue = defaultValue;
        return this;
    }

    public BooleanSchema parseError(String parseError) {
        this.parseError = parseError;
        return this;
    }

    private Boolean coerceType(Object value) {
        if (value instanceof String) {
            return switch ((String) value) {
                case "true", "1", "yes", "on", "y" -> true;
                case "false", "0", "no", "off", "n" -> false;
                default -> null;
            };
        }
        if (value instanceof Boolean) {
            return (Boolean) value;
        }
        if (value instanceof Number) {
            return ((Number) value).intValue() != 0;
        }

        return null;
    }

    @Override
    public ParseResult<Boolean> parse(Object value) {
        if (value == null) {
            if (defaultValue != null) {
                return ParseResult.success(defaultValue);
            }
            if (this.isOptional()) {
                return ParseResult.success(null);
            } else {
                return ParseResult.failure(List.of(parseError));
            }
        }

        if (this.isCoerce()) {
            Boolean coercedValue = coerceType(value);
            if (coercedValue != null) {
                return ParseResult.success(coercedValue);
            }
        }

        if (value instanceof Boolean) {
            return ParseResult.success((Boolean) value);
        } else {
            return ParseResult.failure(List.of(parseError));
        }
    }
}

File: src/main/kotlin/core/BooleanSchema.kt
package studio.styx.sx.core

import studio.styx.sx.types.Schema

class BooleanSchema(
    private val errorMessage: String? = null,
    private val coerce: Boolean = false,
) : Schema<Boolean> {
    private var defaultValue: Boolean? = null
    private var isOptional: Boolean = false
    private var mustBeTrue: Boolean? = null  // null = qualquer, true = só true, false = só false
    private var trueErrorMessage: String? = null
    private var falseErrorMessage: String? = null

    override fun safeParse(input: Any?): SafeParseResult<Boolean> {
        val errors = mutableListOf<String>()
        var parsed: Boolean? = null

        // 1. Null / vazio / optional / default
        if (input == null || input == "") {
            if (defaultValue != null) {
                parsed = defaultValue
            } else if (isOptional) {
                return SafeParseResult.success(null)
            } else {
                errors.add(errorMessage ?: "Value is required")
                return SafeParseResult.failure(errors)
            }
        } else {
            // 2. Coerce ou strict
            parsed = if (coerce) {
                when (input) {
                    is Boolean -> input
                    is Number -> input.toDouble() == 1.0
                    is String -> when (input.trim().lowercase()) {
                        "true", "1", "yes", "on", "t" -> true
                        "false", "0", "no", "off", "f" -> false
                        else -> null
                    }
                    else -> null
                }
            } else {
                input as? Boolean
            }

            if (parsed == null) {
                errors.add(errorMessage ?: "Expected a boolean, got ${input::class.simpleName}")
                return SafeParseResult.failure(errors)
            }
        }

        // 3. Refinamentos .true() / .false()
        val value = parsed!!
        mustBeTrue?.let { expected ->
            if (value != expected) {
                errors.add(
                    if (expected) trueErrorMessage ?: "Must be true"
                    else falseErrorMessage ?: "Must be false"
                )
            }
        }

        return if (errors.isEmpty()) {
            SafeParseResult.success(value)
        } else {
            SafeParseResult.failure(errors)
        }
    }

    // parse / parseOptional (usando safeParse)
    override fun parse(input: Any?): Boolean = safeParse(input).let { r ->
        if (!r.success) throw IllegalArgumentException(r.errors.joinToString(" | "))
        r.value!!
    }

    override fun parseOptional(input: Any?): Boolean? = safeParse(input).value

    // Builders
    fun default(value: Boolean) = apply { defaultValue = value }
    fun optional() = apply { isOptional = true }

    fun isTrue(errorMessage: String? = null) = apply {
        mustBeTrue = true
        trueErrorMessage = errorMessage
    }

    fun isFalse(errorMessage: String? = null) = apply {
        mustBeTrue = false
        falseErrorMessage = errorMessage
    }
}